nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+package nameserver
+
+import (
+	"errors"
+	"github.com/miekg/dns"
+	"math"
+	"sort"
+	"sync"
+	"time"
+	//"fmt"
+	"math/rand"
+)
+
+var (
+	errInvalidCapacity = errors.New("Invalid cache capacity")
+	errCouldNotResolve = errors.New("Could not resolve")
+	errTimeout         = errors.New("Timeout while waiting for resolution")
+)
+
+const (
+	defPendingTimeout int = 5 // timeout for a resolution
+)
+
+type entryStatus uint8
+
+const (
+	stPending  entryStatus = iota // someone is waiting for the resolution
+	stResolved entryStatus = iota // resolved
+)
+
+const (
+	CacheLocalReply uint8 = 1 << iota // the reply was obtained from a local resolution
```

Comment:
That is a procrastination argument :)

In a case like this, I'm fond of flags because _1)_ they look better when calling the function (eg, I think `s.cache.Put(..., CacheLocalReply, ...)` is more intuitive than `s.cache.Put(..., true, ...)`)  _2)_ they do not create functions with a one zillion arguments _3)_ there is no storage/performance gain in using a bool _4)_ and because, in conclusion, there is nothing bad in using a flag and, at the same time, you do not break the API when you add more things... :)